src/graphemebreak.c

```diff
@@ -0,0 +1,252 @@
+#include "graphemebreakdata.c"
+#include "stdbool.h"
+#include "string.h"
+
+#define ARRAY_LEN(x) (sizeof(x) / sizeof(x[0]))
+
+/**
+ * Initializes the wordbreak internals.  It currently does nothing, but
+ * it may in the future.
+ */
+void init_graphemebreak(void)
+{
+}
+
+/**
+ * Gets the grapheme breaking class of a character.
+ *
+ * @param ch   character to check
+ * @return     the grapheme breaking class if found; \c GBP_Other otherwise
+ */
+static enum GraphemeBreakClass get_char_gb_class(utf32_t ch)
+{
+    int min = 0;
+    int max = ARRAY_LEN(gb_prop_default) - 1;
+    int mid;
+
+    do
+    {
+        mid = (min + max) / 2;
+
+        if (ch < gb_prop_default[mid].start)
+            max = mid - 1;
+        else if (ch > gb_prop_default[mid].end)
+            min = mid + 1;
+        else
+            return gb_prop_default[mid].prop;
+    }
+    while (min <= max);
+
+    return GBP_Other;
+}
+
+/**
+ * Sets the grapheme breaking information for a generic input string.
+ *
+ * @param[in]  s             input string
+ * @param[in]  len           length of the input
+ * @param[out] brks          pointer to the output breaking data, containing
+ *                           #GRAPHEMEBREAK_BREAK or #GRAPHEMEBREAK_NOBREAK
+ * @param[in] get_next_char  function to get the next UTF-32 character
+ */
+static void set_graphemebreaks(
+        const void *s,
+        size_t len,
+        char *brks,
+        get_next_char_t get_next_char)
+{
+    size_t posNext = 0;
+    bool rule10Left = false; // is the left side of rule 10 fulfilled?
+    bool evenRegionalIndicators = true; // is the number of preceeding GBP_RegionalIndicator characters even
+
+    utf32_t ch = get_next_char(s, len, &posNext);
+    enum GraphemeBreakClass current_class = get_char_gb_class(ch);
+
+    // initialize whole output to inside char
+    memset(brks, GRAPHEMEBREAK_INSIDEACHAR, len);
+
+    while (true)
+    {
+        enum GraphemeBreakClass prev_class = current_class;
+
+        // safe position if current character so that we can store the result there later on
+        size_t brksPos = posNext-1;
+
+        // get nect character
+        ch = get_next_char(s, len, &posNext);
+
+        if (ch == EOS)
+        {
+          // done, place one final break after the last character as per algorithm rule GB1
+          brks[brksPos] = GRAPHEMEBREAK_BREAK;
+          break;
+        }
+
+        // get class of current character
+        current_class = get_char_gb_class(ch);
+
+        // update some helper variables
+        if ((prev_class == GBP_E_Base) || (prev_class == GBP_E_Base_GAZ))
+          rule10Left = true;
+        else if (prev_class != GBP_Extend)
+          rule10Left = false;
+
+        if (prev_class == GBP_Regional_Indicator)
+          evenRegionalIndicators = !evenRegionalIndicators;
+        else
+          evenRegionalIndicators = true;
+
+        // check all rules
+        if (prev_class == GBP_CR && current_class == GBP_LF)
```

Comment:
I still really don't like the lack of {} for the content of the conditionals, this has only lead to issues in C code.
Goto fail is one example, another is code like:

```
if (a)
    if (b)
        foo()
else
    bar()
```

Which is obviously wrong. Better have curlys and be safe.